### PR TITLE
Якимов Илья. Технология SEQ-MPI. Нахождение максимальных значений по строкам матрицы. Вариант 15.

### PR DESCRIPTION
**Pull request by baifffboy**
## Описание

Алгоритм: Реализован алгоритм поиска максимальных значений в строках матрицы с последующим суммированием результатов. Для каждой строки матрицы находится максимальный элемент, затем вычисляется сумма всех найденных максимумов.

Применение технологии параллелизма (MPI) - архитектура:
1. Мастер-процесс (rank 0): чтение матрицы из файла, распределение данных, сбор результатов
2. Рабочие процессы (rank 1..N-1): обработка выделенных блоков строк

- **Задача**: Нахождение максимальных значений по строкам матрицы.
- **Вариант**: 15
- **Технология**: SEQ, MPI
- **Описание** 

Реализация: Разработана MPI-версия алгоритма поиска максимальных значений в строках матрицы с последующим суммированием результатов. Реализация включает блочное распределение строк матрицы между процессами, эффективную коммуникацию через точечные сообщения и сбор результатов на мастер-процессе.

Отчет содержит:
- Последовательный и параллельный алгоритмы с псевдокодом
- Схему распределения данных и коммуникации MPI
- Результаты тестирования корректности и производительности
- Анализ эффективности при разном количестве процессов
- Выводы о масштабируемости и ограничениях алгоритма
---

## Чек-лист

- [x] **Статус CI**: Все CI-задачи (сборка, тесты, генерация отчёта) успешно проходят на моей ветке в моем форке
- [x] **Директория и именование задачи**: Я создал директорию с именем `<фамилия>_<первая_буква_имени>_<короткое_название_задачи>`
- [x] **Полное описание задачи**: Я предоставил полное описание задачи в теле pull request
- [x] **clang-format**: Мои изменения успешно проходят `clang-format` локально в моем форке (нет ошибок форматирования)
- [x] **clang-tidy**: Мои изменения успешно проходят `clang-tidy` локально в моем форке (нет предупреждений/ошибок)
- [x] **Функциональные тесты**: Все функциональные тесты успешно проходят локально на моей машине
- [x] **Тесты производительности**: Все тесты производительности успешно проходят локально на моей машине
- [x] **Ветка**: Я работаю в ветке, названной точно так же, как директория моей задачи (например, `nesterov_a_vector_sum`), а не в `master`
- [x] **Правдивое содержание**: Я подтверждаю, что все сведения, указанные в этом pull request, являются точными и достоверными

### CODE REVIEW

File: tasks/CMakeLists.txt [hunk 1]
```diff
@@ -1,29 +1,29 @@
...
-  GLOB subdirs
-  RELATIVE "${CMAKE_CURRENT_SOURCE_DIR}"
-  "${CMAKE_CURRENT_SOURCE_DIR}/*")
-foreach(sub IN LISTS subdirs)
-  if(IS_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/${sub}" AND NOT sub STREQUAL
-                                                           "common")
-    ppc_configure_subproject(${sub})
-  endif()
-endforeach()
+project(parallel_programming_course LANGUAGES C CXX)
```
Comment (allnes) [inline tasks/CMakeLists.txt]: Для новой задачи нет info.json/settings.json и явной регистрации в CMake/scoreboard, поэтому генераторы задач и сборка не узнают о её существовании. Добавьте метаданные в каталог задачи, подключите их в tasks/CMakeLists и убедитесь, что scoreboard/CI могут подхватить и прогнать тесты для этой задачи.

File: tasks/yakimov_i_max_values_in_matrix_rows/seq/src/ops_seq.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  // std::filesystem::path exe_path = std::filesystem::read_symlink("/proc/self/exe");
+  std::filesystem::path base_path = std::filesystem::current_path();
+  while (base_path.filename() == "build") {
+    base_path = base_path.parent_path();
+  }
+  matrix_Filename_ =
+      base_path.string() + "/tasks/yakimov_i_max_values_in_matrix_rows/data/" + std::to_string(GetInput()) + ".txt";
+}
+
+bool YakimovIMaxValuesInMatrixRowsSEQ::ValidationImpl() {  // NOLINT(readability-convert-member-functions-to-static)
```
Comment (aobolensk) [inline tasks/yakimov_i_max_values_in_matrix_rows/seq/src/ops_seq.cpp]: Please, remove all NOLINTs

**Comment by aobolensk** [comment]
Revert unrelated changes

File: tasks/yakimov_i_max_values_in_matrix_rows/data/d2utmptbLvC3
Comment (aobolensk) [inline tasks/yakimov_i_max_values_in_matrix_rows/data/d2utmptbLvC3]: Please, do not push binaries

File: tasks/yakimov_i_max_values_in_matrix_rows/data/generate_matrices.cpp
Comment (aobolensk) [inline tasks/yakimov_i_max_values_in_matrix_rows/data/generate_matrices.cpp]: I guess this thing is needed for you, but actually not required to be uploaded

File: tasks/yakimov_i_max_values_in_matrix_rows/tests/functional/functional.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  }
+};
+
+namespace {
+
+TEST_P(YakimovIMaxValuesInMatrixRowsFuncTests, MaxValuesInRows) {
+  ExecuteTest(GetParam());
+}
+
+const std::array<TestType, 5> kTestParam = {std::make_tuple(1, "tiny"), std::make_tuple(2, "very_small"),
```
Comment (aobolensk) [inline tasks/yakimov_i_max_values_in_matrix_rows/tests/functional/functional.cpp]: Please, add more tests. You don't even use all files you have uploaded. Also consider checking corner cases

File: tasks/yakimov_i_max_values_in_matrix_rows/tests/performance/performance.cpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ protected:
+  bool CheckTestOutputData(OutType &output_data) final {
+    return output_data > 0;
+  }
+
+  InType GetTestInputData() final {
+    return 30;
+  }
+};
+
```
[the remark this answers is not in the document]
Comment (baifffboy) [reply tasks/yakimov_i_max_values_in_matrix_rows/tests/performance/performance.cpp]: add more perfomance tests

**Review by aobolensk — APPROVED** [review]
